### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by alexodan**
## Summary

This PR has the basic auth mechanisms. A login page (under `/`) a register (`/register`) and account pages (`/account`).
I chose supabase to handle users (and later the chats). Everything needed in the `.env.local` is in the private repository.

## Learnings

- When using a vpn you might not get the email on signin with magic link 💀 (this was using next-auth tho).
- Use the library you choose for the job, not fight it. And RTFriendlyM! (see 2nd point in struggles). Supabase has methods for signing in, up and out. I was trying to do that manually first then looking again I found it was solved out of the box.

## Struggles

- About Next.js, and I'm still getting used to the whole server/client approach in the same place. Sometimes I wonder if a file will run server or client side, or at build time or run time, etc. The errors most of the time are helpful though, and the community has most of the answers.
- Trying to set cookies/handle auth manually. Looking back I think I was just _not_ using the library I chose for the task (supabase).
- SSR and hydration. Using styled components apparently requires a pretty good understanding of these two, for now I think I'll opt for not styling that way.

## References

- https://supabase.com/docs/guides/getting-started/tutorials/with-react

### CODE REVIEW

File: styled-system/helpers.mjs [hunk 1]
```diff
@@ -0,0 +1,280 @@
+// src/assert.ts
```
Comment (sandrina-p) [inline styled-system/helpers.mjs:1]: question: I confess I'm surprised the entire `styled-system` is here in the codebase. Is this really intended?

Comment (sandrina-p) [inline styled-system/helpers.mjs]: question: I confess I'm surprised the entire `styled-system` is here in the codebase. Is this really intended?

File: .prettierrc [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  "arrowParens": "avoid",
+  "semi": false,
+  "singleQuote": false,
```
Comment (sandrina-p) [inline .prettierrc]: question: In the other projects you used single quote, why not in this one? 

(I confess I prefer single quote too eheh)

File: .prettierrc [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+  "trailingComma": "es5"
```
Comment (sandrina-p) [inline .prettierrc]: question: In the other project you always had trailingComma (`all`) but not here. Why?

An advantage of `all`, is that it makes the git diff smaller. Without it, 2 lines are changed: (the new one added and the one before with the new trailing) With it, just one line is changed: (the new one added)

File: src/app/account/account-form.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
+"use client"
+
+import { useCallback, useEffect, useState } from "react"
+import { Database } from "../../types/supabase"
```
Comment (sandrina-p) [inline src/app/account/account-form.tsx]: suggestion: You could use [module alias](https://nextjs.org/docs/app/building-your-application/configuring/absolute-imports-and-module-aliases#module-aliases) to make your imports much easier across the codebase ;)

File: src/app/login-form.tsx [hunk 5]
```diff
@@ -0,0 +1,140 @@
+"use client"
```
Comment (sandrina-p) [inline src/app/login-form.tsx]: suggestion: You created a folder for every route, except the root. I found it odd having this here, perhaps you could look at [Next colocation recommendation](https://nextjs.org/docs/app/building-your-application/routing/colocation), and find a better place for this one. For example `app/root/` or `src/components`. 

Another solution totally different is to move this view to `/login` route and reserve `/` for something else, usually, it's used for marketing/landing page. 

Just a suggestion, it's up to you.

File: src/app/login-form.tsx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    return redirect("/messages")
+  }
+
+  return (
+    <>
+      {!isMagicLinkLogin && (
+        <>
+          <form onSubmit={handleSubmit}>
+            <div>
+              <label htmlFor="email-login">
```
Comment (sandrina-p) [inline src/app/login-form.tsx]: question/todo: Why do you have a htmlFor here? How do you know if you need it or not? [resource](https://css-tricks.com/html-inputs-and-labels-a-love-story/).

File: src/app/login-form.tsx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            }}
+            localization={{
+              variables: {
+                magic_link: {
+                  email_input_label: "Login with magic link",
+                },
+              },
+            }}
+          />
+          <Separator text="OR" userCss={css({ mb: 2 })} />
```
Comment (sandrina-p) [inline src/app/login-form.tsx]: todo: Always avoid caps because of A11Y. This is readable by a screen as an acronym "O.R." not as "or". Use CSS to make it all caps ;)

File: src/app/login-form.tsx [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  const [isMagicLinkLogin, setIsMagicLinkLogin] = useState(false)
+
+  const handleSubmit = async (e: React.FormEvent) => {
+    e.preventDefault()
+    const { error } = await supabase.auth.signInWithPassword({
+      email,
+      password,
+    })
+    if (error) {
+      console.error(error)
```
Comment (sandrina-p) [inline src/app/login-form.tsx]: todo: Can't you do better to show the error in the UI? :p

File: src/app/register/page.tsx [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+export default async function Register() {
+  const supabase = createServerComponentClient<Database>({ cookies })
+
+  const {
+    data: { session },
+  } = await supabase.auth.getSession()
+
+  if (session) {
+    return redirect("/")
+  }
```
Comment (sandrina-p) [inline src/app/register/page.tsx:16]: todo: Have you ever used a HoC for this redirect? It's perfect to avoid repeating this in every route. [example](https://louisrli.github.io/blog/2019/11/17/auth-redirect-next-js/)

File: src/app/register/register-form.tsx [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setFormData({ ...formData, [e.target.name]: e.target.value })
+  }
+
+  const handleSubmit = async (e: React.FormEvent) => {
+    e.preventDefault()
+    if (password !== confirmPassword) {
+      alert("Passwords don't match")
+    } else {
+      const { error } = await supabase.auth.signUp({
```
Comment (sandrina-p) [inline src/app/register/register-form.tsx:37]: bug: On my side, when I login, the request succeded but I'm not redirected to any page. I believe it's missing the redirect after the response.

File: src/app/login-form.tsx [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+  const [password, setPassword] = useState("")
+  const [isMagicLinkLogin, setIsMagicLinkLogin] = useState(false)
+
+  const handleSubmit = async (e: React.FormEvent) => {
+    e.preventDefault()
+    const { error } = await supabase.auth.signInWithPassword({
+      email,
+      password,
+    })
+    if (error) {
```
Comment (sandrina-p) [inline src/app/login-form.tsx]: bug: I can't login with credentials, only with the magic link. Is it the same for you? 🤔

File: src/app/login-form.tsx [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        </Button>
+      )}
+      {isMagicLinkLogin && (
+        <>
+          <Auth
+            supabaseClient={supabase}
+            view="magic_link"
+            showLinks={false}
+            providers={[]}
+            redirectTo="http://localhost:3000/auth/callback"
```
Comment (sandrina-p) [inline src/app/login-form.tsx]: bug: If i use a magic link twice, I get redirected to the homepage without any feedback, even thought the URL is: 

```
http://localhost:3000/#error=unauthorized_client&error_code=401&error_description=Email+link+is+invalid+or+has+expired
```

Handling these errors from the URL would be nice too. It can be now or in another MR, as you wish.

File: src/app/account/account-form.tsx [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+      }
+
+      if (data) {
+        setFullName(data.full_name)
+        setUsername(data.username)
+        setWebsite(data.website)
+        setAvatarUrl(data.avatar_url)
+      }
+    } catch (error) {
+      alert("Error loading user data!")
```
Comment (sandrina-p) [inline src/app/account/account-form.tsx]: bug: I always get an error here. Is it expected?

```js
{
    "code": "42703",
    "details": null,
    "hint": null,
    "message": "column \"avatarUrl\" of relation \"profiles\" does not exist"
}
```

**Review by sandrina-p — CHANGES_REQUESTED** [review]

File: middleware.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
+import { createMiddlewareClient } from "@supabase/auth-helpers-nextjs"
```
Comment (sandrina-p) [inline middleware.ts]: > About Next.js, and I'm still getting used to the whole server/client approach in the same place. Sometimes I wonder if a file will run server or client side, or at build time or run time, etc. The errors most of the time are helpful though, and the community has most of the answers.

Based on its docs, every file under `/app` is considered a SSC unless you use `use client`. [resource](https://nextjs.org/docs/app/building-your-application/routing#the-app-router). I also liked this [When to use server and client components](https://nextjs.org/docs/getting-started/react-essentials#when-to-use-server-and-client-components) diagram (and the whole page in general)

File: styled-system/helpers.mjs (hunk 1 above)
Comment (alexodan) [reply styled-system/helpers.mjs]: apparently not 😆 https://panda-css.com/docs/overview/faq#should-i-commit-the-styled-system-folder (the answer is no it shouldn't be committed). will remove it

File: .prettierrc (hunk 2 above)
Comment (alexodan) [reply .prettierrc]: haha I enabled copilot and started hitting tab, prettier is working funny in my IDE now but yeah I might change it haha

File: .prettierrc (hunk 3 above)
Comment (alexodan) [reply .prettierrc]: oh good one, I'll update it

File: .prettierrc (hunk 2 above)
Comment (sandrina-p) [reply .prettierrc]: I see. Remember, you are the pilot, and as the pilot you are the one making the final decisions. Don't get blindfolded by a robot _copilot_. ;)

File: src/app/account/account-form.tsx (hunk 13 above)
Comment (alexodan) [reply src/app/account/account-form.tsx]: I think it's because I haven't protected that route, need to do that cuz if u're not logged in it will throw an error

File: src/app/login-form.tsx (hunk 12 above)
Comment (alexodan) [reply src/app/login-form.tsx]: interesting, not sure how to do that but I'll investigate

File: src/app/login-form.tsx (hunk 11 above)
Comment (alexodan) [reply src/app/login-form.tsx]: I am able to login hehe, maybe we can try on our call and see what's going on

File: src/app/login-form.tsx (hunk 6 above)
Comment (alexodan) [reply src/app/login-form.tsx]: I forgot the `id` in the input. After reading part of the article I changed my mind and extracted the input outside the label and added a text inside it.

File: src/app/login-form.tsx (hunk 7 above)
Comment (alexodan) [reply src/app/login-form.tsx]: TIL! hehe

File: src/app/register/page.tsx (hunk 9 above)
Comment (alexodan) [reply src/app/register/page.tsx:16]: I realize now that I needed this cuz the middleware wasn't working. It's working now so I removed this condition (problem was that `middleware.ts` needs to be under `src`, I had it under the root dir 🤪)

File: src/app/account/account-form.tsx (hunk 13 above)
Comment (alexodan) [reply src/app/account/account-form.tsx]: this should be solved now that the middleware works
